### PR TITLE
fix windows related encoding problems

### LINKED ISSUE #983: fluidsynth for windows failed to load soundfont path containing non-ASCII chars
### FluidSynth 2.2.3
os: win10
locale(ANSI encoding): CP936(default for widows for china since windows xp?)

### Describe the bug
```cmd
d:\opt\fluidsynth-2.2.3\bin>fluidsynfluidsynth.exe d:\冷备\音乐制作\软件\soundfont2\GeneralUser-GS-Live-Audigy-v1.43.sf2
FluidSynth runtime version 2.2.3
Copyright (C) 2000-2021 Peter Hanappe and others.
Distributed under the LGPL license.
SoundFont(R) is a registered trademark of Creative Technology Ltd.

fluidsynth: error: fluid_is_soundfont(): fopen() failed: 'File does not exist.'
Parameter 'd:\冷备\音乐制作\软件\soundfont2\GeneralUser-GS-Live-Audigy-v1.43.sf2' not a SoundFont or MIDI file or error occurred identifying it.
```
But that file indeed exists.

### Expected behavior
Soundfont file should be loaded.

### Additional context
seems related to https://github.com/FluidSynth/fluidsynth/issues/880

### PR DESCRIPTION
**Pull request by tsingakbar**
Resolves #983

### CODE REVIEW

File: src/fluidsynth.c [hunk 1]
```diff
@@ -896,14 +896,25 @@ int main(int argc, char **argv)
     /* load the soundfonts (check that all non options are SoundFont or MIDI files) */
     for(i = arg1; i < argc; i++)
     {
-        if(fluid_is_midifile(argv[i]))
+#if defined(WIN32)
+        int u16_count = MultiByteToWideChar(CP_ACP, 0, argv[i], -1, NULL, 0);
+        char *u16_buf = malloc((size_t)u16_count * 2);
```
Comment (derselbst) [inline src/fluidsynth.c:901]: Pls. use something like
```suggestion
        LPWSTR u16_buf = FLUID_ARRAY(*LPWSTR, u16_count);
```

Comment (derselbst) [reply src/fluidsynth.c:901]: And pls. check for NULL. Same below.

File: src/fluidsynth.c [hunk 2]
```diff
@@ -896,14 +896,25 @@ int main(int argc, char **argv)
...
+        u16_count = MultiByteToWideChar(CP_ACP, 0, argv[i], -1, (LPWSTR)u16_buf, u16_count);
+        int u8_byte_count = WideCharToMultiByte(CP_UTF8, 0, (LPWSTR)u16_buf, u16_count, NULL, 0, NULL, NULL);
+        char *u8_path = malloc(u8_byte_count);
+        u8_byte_count = WideCharToMultiByte(CP_UTF8, 0, (LPWSTR)u16_buf, u16_count, u8_path, u8_byte_count, NULL, NULL);
+        free(u16_buf);
```
Comment (derselbst) [inline src/fluidsynth.c:906]: Pls use
```suggestion
        FLUID_FREE(u16_buf);
```

Comment (tsingakbar) [reply src/fluidsynth.c]: all fixed.
Allso tried to use FLUID_MALLOC to in pair with FLUID_FREE, but found it not exported later. So currently malloc and FLUID_FREE is used in the fluidsynth executable.

**Review by derselbst — COMMENTED** [review]
Sorry, there are still a few issues: if malloc returns NULL, pls exit the program by:

```
fprintf(stderr, "Out of memory\n");
return -1;
```

File: src/fluidsynth.c [hunk 3]
```diff
@@ -896,14 +896,33 @@ int main(int argc, char **argv)
...
+        const char *u8_path = argv[i];
+#if defined(WIN32)
+        /* try to convert ANSI encoding path to UTF8 encoding path */
+        char *u8_buf = NULL;
+        int u16_count = MultiByteToWideChar(CP_ACP, 0, argv[i], -1, NULL, 0);
+        LPWSTR u16_buf = (LPWSTR)malloc((size_t)u16_count * 2);
+        if (u16_buf != NULL)
+        {
+            u16_count = MultiByteToWideChar(CP_ACP, 0, argv[i], -1, u16_buf, u16_count);
+            int u8_byte_count = WideCharToMultiByte(CP_UTF8, 0, u16_buf, u16_count, NULL, 0, NULL, NULL);
```
[the remark this answers is not in the document]
Comment (tsingakbar) [reply src/fluidsynth.c:908]: ok. I wrapped these codes to a function, and make sure all variables declared at the beginning of scopes.

[the remark this answers is not in the document]
Comment (tsingakbar) [reply src/fluidsynth.c:908]: For the malloc failure, I will return NULL from the function, and main() will continue trying to use argv[i] as u8_path. If this is still not acceptable, I will continue to modify.

File: src/fluidsynth.c [hunk 4]
```diff
@@ -896,14 +896,33 @@ int main(int argc, char **argv)
     /* load the soundfonts (check that all non options are SoundFont or MIDI files) */
     for(i = arg1; i < argc; i++)
     {
-        if(fluid_is_midifile(argv[i]))
+        const char *u8_path = argv[i];
+#if defined(WIN32)
+        /* try to convert ANSI encoding path to UTF8 encoding path */
+        char *u8_buf = NULL;
+        int u16_count = MultiByteToWideChar(CP_ACP, 0, argv[i], -1, NULL, 0);
+        LPWSTR u16_buf = (LPWSTR)malloc((size_t)u16_count * 2);
```
[the remark this answers is not in the document]
Comment (tsingakbar) [reply src/fluidsynth.c:904]: ok. `sizeof(*LPWSTR)` is invalid at least for msvc(vs2019), so I will directly use WHCAR type.

**Review by derselbst — APPROVED** [review]
Your latest change didn't compile on non-windows platforms. I made some adjustments, now it should be good.

Also, I'm now exiting the program if `win32_ansi_to_utf8` fails. Silently continuing when an error occurred is not good practice.

Anyway, thanks again!

**Pull request merged**